### PR TITLE
Total counts - Distributed Tokens query and ah volume query

### PR DESCRIPTION
**Pull request by alchemistgo87**

### CODE REVIEW

File: crates/core/src/db/models.rs [hunk 1]
```diff
@@ -3499,3 +3499,19 @@ pub struct DolphinStats1D<'a> {
     pub last_listed_1d: i64,
     pub last_volume_1d: i64,
 }
+
+/// A row in a `tokens_distributed` query, representing total $SAUCE tokens distributed between particular dates.
+#[derive(Debug, Clone, Queryable, QueryableByName)]
+pub struct TokensDistributed {
+    /// Total tokens distributed
+    #[sql_type = "Numeric"]
+    pub tokens_distributed: BigDecimal,
```
Comment (kespinola) [inline crates/core/src/db/models.rs:3508]: This should be big_int won't be any decimals involved.

File: crates/core/src/db/models.rs [hunk 2]
```diff
@@ -3499,3 +3499,19 @@ pub struct DolphinStats1D<'a> {
...
+}
+
+/// Auction House Volume
+#[derive(Debug, Clone, Queryable, QueryableByName)]
+pub struct AuctionHouseVolume {
+    /// Auction House Volume
+    #[sql_type = "Numeric"]
+    pub volume: BigDecimal,
```
Comment (kespinola) [inline crates/core/src/db/models.rs:3516]: Same volume is BigInt since its lamports.

File: crates/core/src/db/queries/auction_house.rs [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+/// Load token distributed for reward center.
+///
+/// # Errors
+/// This function fails if the underlying SQL query returns an error
+pub fn volume(
+    conn: &Connection,
+    address: impl ToSql<Text, Pg>,
+    start_date: NaiveDateTime,
+    end_date: NaiveDateTime,
```
Comment (kespinola) [inline crates/core/src/db/queries/auction_house.rs:37]: Start and end data should be optional

File: crates/core/src/db/queries/reward_centers.rs [hunk 4]
```diff
@@ -69,3 +69,33 @@ pub fn payouts(
...
+
+/// Load token distributed for reward center.
+///
+/// # Errors
+/// This function fails if the underlying SQL query returns an error
+pub fn tokens_distributed(
+    conn: &Connection,
+    address: impl ToSql<Text, Pg>,
+    start_date: NaiveDateTime,
+    end_date: NaiveDateTime,
```
Comment (kespinola) [inline crates/core/src/db/queries/reward_centers.rs:93]: start and end date should be optional.

File: crates/core/src/db/queries/auction_house.rs [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+const VOLUME_QUERY: &str = r"
+SELECT COALESCE(SUM(purchase_price), 0) as volume
+    FROM puchases p
+    INNER JOIN auction_houses ah
+        on (p.auction_house = ah.address)
+    WHERE p.auction_house = $1
+    AND p.created_at >= $2
+    AND p.created_at <= $3
+;
```
Comment (kespinola) [inline crates/core/src/db/queries/auction_house.rs:23]: this needs to also support start_date and end_date as optional.

File: crates/graphql/src/schema/objects/auction_house.rs [hunk 6]
```diff
@@ -135,4 +139,30 @@ impl AuctionHouse {
...
+    pub async fn volume(
+        &self,
+        ctx: &AppContext,
+        start_date: Option<DateTime<Utc>>,
+        end_date: Option<DateTime<Utc>>,
+    ) -> FieldResult<Option<U64>> {
+        let conn = ctx.shared.db.get()?;
+
+        let start_date = start_date.unwrap_or_default();
+        let end_date = end_date.unwrap_or(Utc::now());
```
Comment (kespinola) [inline crates/graphql/src/schema/objects/auction_house.rs:152]: None should be supported.

File: crates/graphql/src/schema/objects/reward_center.rs [hunk 7]
```diff
@@ -105,4 +108,27 @@ impl RewardCenter {
...
+        start_date: Option<DateTime<Utc>>,
+        end_date: Option<DateTime<Utc>>,
+    ) -> FieldResult<Option<U64>> {
+        let conn = context.shared.db.get()?;
+
+        let result = queries::reward_centers::tokens_distributed(
+            &conn,
+            &self.address,
+            start_date.map(|v| v.naive_utc()),
+            end_date.map(|v| v.naive_utc()),
```
Comment (kespinola) [inline crates/graphql/src/schema/objects/reward_center.rs:124]: This should be ok to be none.

Comment (alchemistgo87) [reply crates/graphql/src/schema/objects/reward_center.rs:124]: This is already optional.

File: crates/graphql/src/schema/objects/auction_house.rs [hunk 8]
```diff
@@ -135,4 +139,27 @@ impl AuctionHouse {
...
+            &self.address,
+            start_date.map(|v| v.naive_utc()),
+            end_date.map(|v| v.naive_utc()),
+        )?;
+
+        Ok(result
+            .into_iter()
+            .next()
+            .map(|models::AuctionHouseVolume { volume }| {
+                volume.to_u64().unwrap_or_default().into()
```
Comment (kespinola) [inline crates/graphql/src/schema/objects/auction_house.rs:162]: @ray-kast is this correct?

Comment (alchemistgo87) [reply crates/graphql/src/schema/objects/auction_house.rs:162]: @ray-kast Using `volume.to_u64().ok_or_else(|| anyhow!("..."))` throws mismatched types error:
```
mismatched types
expected enum `std::result::Result<std::option::Option<bigint::U64>, juniper::FieldError>`
   found enum `std::result::Result<std::option::Option<u64>, anyhow::Error>`
```

Comment (alchemistgo87) [reply crates/graphql/src/schema/objects/auction_house.rs:162]: @ray-kast That worked. Thanks!

**Review by kespinola — APPROVED** [review]

**Pull request merged**